protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java

```diff
@@ -0,0 +1,305 @@
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.remove(messageUid1);
+
+        assertThat(testee.getMsn(messageUid1).isPresent()).isFalse();
+        assertThat(testee.getMsn(messageUid2).isPresent()).isTrue();
+        assertThat(testee.getMsn(messageUid3).isPresent()).isTrue();
+    }
+
+    @Test
+    public void  removeShouldRemoveOnlyTheCorrectUidWhenOneTheEnd() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.remove(messageUid3);
+
+        assertThat(testee.getMsn(messageUid3).isPresent()).isFalse();
+        assertThat(testee.getMsn(messageUid1).isPresent()).isTrue();
+        assertThat(testee.getMsn(messageUid2).isPresent()).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepMessageNumberContiguousWhenDeletingTheFirstOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid1);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepMessageNumberContiguousWhenDeletingTheLastOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid4);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepMessageNumberContiguousWhenDeletingOneInTheMiddle() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid3);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepUidInOrderWhenDeletingTheFirstOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid1);
+
+        assertThat(uidAreInOrder(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepUidInOrderWhenDeletingTheLastOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid4);
+
+        assertThat(uidAreInOrder(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepUidInOrderWhenDeletingOneInTheMiddle() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid3);
+
+        assertThat(uidAreInOrder(testee)).isTrue();
+    }
+
+    private static boolean messageNumberAreContiguous(UidMsnMapper testee) {
+        int first = UidMsnMapper.FIRST_MSN;
```

Comment:
1: not sure it will make the code more consistent, readable, please propose me code if you find something really more consistent.
2: I use the concept of messageNumberAreContigious at least three time, I think it make test more shorter and more comprehensible than using asserts directly one code.
For @VisibleForTesting, I don't like that because it make you code dependant of the inner implementation data structure which is IMO a bad idea when that can be avoided. And here it can be.